### ISSUE TITLE
How to get error in each batch iteration?

### ISSUE THREAD

**Issue by lincoln2010** [apache/incubator-mxnet#3510]
I am now using mxnet and want to get error in each update in optimizer.py. But can't find any method to get the current error in each iteration, there are no function like def _get_lr() to get the error directly. Thank you very much.

**Comment by RogerBorras** [comment]
Here you are an example in R:


nnet_strain <- mx.symbol.Variable('data') %>%
                 mx.symbol.FullyConnected(name = 'fc1', num_hidden = 5) %>%
                 mx.symbol.Activation(name = 'relu1', act_type ='relu') %>%
                 mx.symbol.FullyConnected(name = 'fc2', num_hidden = 5) %>%
                 mx.symbol.SoftmaxOutput(name = 'softmax')
                 #mx.symbol.LogisticRegressionOutput(name = 'logisticregression')


# mx.callback.plot.train.metric

mx.callback.plot.train.metric <- function(period, logger = NULL){
  function(iteration, nbatch, env, verbose=TRUE){
    if (nbatch %% period == 0 && !is.null(env$metric)){
      N = env$end.round
      result <- env$metric$get(env$train.metric)
      plot(c(0.0, 1) ~ c(0, N), col=NA, ylab=paste0("Train ", result$name), xlab="")
      logger$train <- c(logger$train, result$value)
      lines(logger$train, lwd=3, col="red")
    }
    return(TRUE)
  }
}


devices <- mx.cpu() #mx.ctx.default(mx.cpu(1))

logger <- mx.metric.logger$new()

tic = proc.time()

#mx.set.seed(1)
mx.set.seed(19)
mx.set.seed(12)
model_strain <- mx.model.FeedForward.create(nnet_strain,
                                     X = t(strain[,-c(1)]),
                                     y = strain[,1],
                                     ctx = devices,
                                     num.round = 1500,
                                     array.batch.size = 100,
                                     learning.rate = 0.001,
                                     momentum = 0.9,
                                     eval.metric = mx.metric.accuracy,
                                     initializer = mx.init.uniform(0.01),
                                     epoch.end.callback = mx.callback.plot.train.metric(100, logger))

proc.time() - tic

logger_df <- data.frame(Interation=1:length(logger$train), Accuracy = 100*logger$train, Train_error = 100*(1-logger$train))

rm(logger)

ggplot(logger_df, aes(Interation, Accuracy)) + geom_line(lwd=1, col=I("darkblue")) +
  xlab("ITERATION") + ylab("TRAIN ACCURACY (%)") + theme_economist() + scale_colour_economist() +  ylim(0, 100) +
  theme(axis.text = element_text(size=14, colour=NULL),
        axis.title.x = element_text(face="bold", colour="#990000", size=20),
        axis.title.y = element_text(face="bold", colour="#990000", size=20))

**Comment by RogerBorras** [comment]
Here you are an example in R:

`
nnet_strain <- mx.symbol.Variable('data') %>%
                 mx.symbol.FullyConnected(name = 'fc1', num_hidden = 5) %>%
                 mx.symbol.Activation(name = 'relu1', act_type ='relu') %>%
                 mx.symbol.FullyConnected(name = 'fc2', num_hidden = 5) %>%
                 mx.symbol.SoftmaxOutput(name = 'softmax')
                 #mx.symbol.LogisticRegressionOutput(name = 'logisticregression')


# mx.callback.plot.train.metric

mx.callback.plot.train.metric <- function(period, logger = NULL){
  function(iteration, nbatch, env, verbose=TRUE){
    if (nbatch %% period == 0 && !is.null(env$metric)){
      N = env$end.round
      result <- env$metric$get(env$train.metric)
      plot(c(0.0, 1) ~ c(0, N), col=NA, ylab=paste0("Train ", result$name), xlab="")
      logger$train <- c(logger$train, result$value)
      lines(logger$train, lwd=3, col="red")
    }
    return(TRUE)
  }
}


devices <- mx.cpu() #mx.ctx.default(mx.cpu(1))

logger <- mx.metric.logger$new()

tic = proc.time()

#mx.set.seed(1)
mx.set.seed(19)
mx.set.seed(12)
model_strain <- mx.model.FeedForward.create(nnet_strain,
                                     X = t(strain[,-c(1)]),
                                     y = strain[,1],
                                     ctx = devices,
                                     num.round = 1500,
                                     array.batch.size = 100,
                                     learning.rate = 0.001,
                                     momentum = 0.9,
                                     eval.metric = mx.metric.accuracy,
                                     initializer = mx.init.uniform(0.01),
                                     epoch.end.callback = mx.callback.plot.train.metric(100, logger))

proc.time() - tic

logger_df <- data.frame(Interation=1:length(logger$train), Accuracy = 100*logger$train, Train_error = 100*(1-logger$train))

rm(logger)

ggplot(logger_df, aes(Interation, Accuracy)) + geom_line(lwd=1, col=I("darkblue")) +
  xlab("ITERATION") + ylab("TRAIN ACCURACY (%)") + theme_economist() + scale_colour_economist() +  ylim(0, 100) +
  theme(axis.text = element_text(size=14, colour=NULL),
        axis.title.x = element_text(face="bold", colour="#990000", size=20),
        axis.title.y = element_text(face="bold", colour="#990000", size=20))
`

**Comment by RogerBorras** [comment]
Here you are an example with R (mx.callback.plot.train.metric):

`
nnet_strain <- mx.symbol.Variable('data') %>%
                 mx.symbol.FullyConnected(name = 'fc1', num_hidden = 5) %>%
                 mx.symbol.Activation(name = 'relu1', act_type ='relu') %>%
                 mx.symbol.FullyConnected(name = 'fc2', num_hidden = 5) %>%
                 mx.symbol.SoftmaxOutput(name = 'softmax')
                 #mx.symbol.LogisticRegressionOutput(name = 'logisticregression')


### mx.callback.plot.train.metric

mx.callback.plot.train.metric <- function(period, logger = NULL){
  function(iteration, nbatch, env, verbose=TRUE){
    if (nbatch %% period == 0 && !is.null(env$metric)){
      N = env$end.round
      result <- env$metric$get(env$train.metric)
      plot(c(0.0, 1) ~ c(0, N), col=NA, ylab=paste0("Train ", result$name), xlab="")
      logger$train <- c(logger$train, result$value)
      lines(logger$train, lwd=3, col="red")
    }
    return(TRUE)
  }
}


devices <- mx.cpu() #mx.ctx.default(mx.cpu(1))

logger <- mx.metric.logger$new()

tic = proc.time()

mx.set.seed(1)

model_strain <- mx.model.FeedForward.create(nnet_strain,
                                     X = t(strain[,-c(1)]),
                                     y = strain[,1],
                                     ctx = devices,
                                     num.round = 1500,
                                     array.batch.size = 100,
                                     learning.rate = 0.001,
                                     momentum = 0.9,
                                     eval.metric = mx.metric.accuracy,
                                     initializer = mx.init.uniform(0.01),
                                     epoch.end.callback = mx.callback.plot.train.metric(100, logger))

proc.time() - tic

logger_df <- data.frame(Interation=1:length(logger$train), Accuracy = 100*logger$train, Train_error = 100*(1-logger$train))

rm(logger)

ggplot(logger_df, aes(Interation, Accuracy)) + geom_line(lwd=1, col=I("darkblue")) +
  xlab("ITERATION") + ylab("TRAIN ACCURACY (%)") + theme_economist() + scale_colour_economist() +  ylim(0, 100) +
  theme(axis.text = element_text(size=14, colour=NULL),
        axis.title.x = element_text(face="bold", colour="#990000", size=20),
        axis.title.y = element_text(face="bold", colour="#990000", size=20))

`

**Comment by RogerBorras** [comment]
And this code produce this figure:

[SESSIO_DEEP_LEARNING_UPF.pdf](https://github.com/dmlc/mxnet/files/526952/SESSIO_DEEP_LEARNING_UPF.pdf)

**Comment by shuokay** [comment]
You can define your own `EvalMetric`. Refer to `mxnet/python/mxnet/metric.py`. The example below may help you.
https://github.com/xlvector/learning-dl/blob/master/mxnet/triple-loss/triplet_loss.py#L134

**Comment by lincoln2010** [comment]
Thank you very much for your reply

**Comment by lincoln2010** [comment]
Sorry I might write it clearly. The problem is that how to get the error in each batch iteration and get the value of error in optimizer.py because we would like to change the optimization algorithm using the error. Many thanks in advance.

**Comment by lincoln2010** [comment]
PS: Now from the interface in optimizer.py, we can only get the gradient value but not the error value.

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
